File: server/internal/loadout/loadout.go
// Package loadout owns the equipped set: which content fills which slot, what
// arrangement is legal, and how a saved arrangement is carried onto today's
// content. It holds no world state, so the simulation and the store agree on
// one answer without either owning the rule.
//
// The economy's keystone rule — a loadout may only be changed inside a safe
// zone — is not here, because it needs a position. `game.World.SetLoadout`
// enforces it; everything about *what* is a legal set lives in this package.
package loadout

import (
	"fmt"
	"sort"
	"strings"

	"spellfire/server/internal/crafting"
	"spellfire/server/internal/model"
	"spellfire/server/internal/progression"
	"spellfire/server/internal/tuning"
)

// Slot kinds. A Gunslinger's bar is one weapon followed by gadget slots; a
// Mage's is spell slots, with its staff cast through whichever spell is
// selected. Both lay out to the same width, which the tables validate.
const (
	KindWeapon   = "weapon"
	KindGadget   = "gadget"
	KindSpell    = "spell"
	KindKeystone = "keystone"
)

// Slot is one selectable position on the action bar, resolved against the
// tables. An empty slot carries its kind and nothing else.
type Slot struct {
	Index     int
	Kind      string
	ID        string
	Name      string
	AbilityID string
	Element   string
	// Item is the crafted instance a weapon slot holds, and is empty when the
	// slot holds a stock weapon row. It is what the simulation applies component
	// modifiers from.
	Item model.CraftedItem
}

// Filled reports whether the slot holds content that can be used.
func (s Slot) Filled() bool { return s.ID != "" }

// Bar lays an equipped set out as the selectable action bar, in binding order.
// It is the single answer to "what does the use button do", shared by the
// simulation and the menu.
func Bar(tables *tuning.Tables, class model.Class, inventory crafting.Inventory, set model.Loadout) []Slot {
	table := tables.Loadout
	slots := make([]Slot, 0, table.BarSlots())
	if class == model.Gunslinger {
		slots = append(slots, weaponSlot(tables, inventory, 0, set.Weapon))
		for index := 0; index < table.GadgetSlots; index++ {
			slots = append(slots, gadgetSlot(tables, len(slots), at(set.Gadgets, index)))
		}
		return slots
	}
	staff, item, _ := inventory.Equipped(tables, set.Weapon)
	for index := 0; index < table.SpellSlots; index++ {
		slots = append(slots, spellSlot(tables, staff, item, index, at(set.Spells, index)))
	}
	return slots
}

// weaponSlot resolves the one weapon slot. The reference is either a stock
// weapon row or a crafted instance of one; both answer with the row's ability,
// and the instance rides along so the simulation can apply its components.
func weaponSlot(tables *tuning.Tables, inventory crafting.Inventory, index int, id string) Slot {
	weapon, item, ok := inventory.Equipped(tables, id)
	if !ok {
		return Slot{Index: index, Kind: KindWeapon}
	}
	return Slot{Index: index, Kind: KindWeapon, ID: id, Name: weapon.Name, AbilityID: weapon.Ability, Item: item}
}

func gadgetSlot(tables *tuning.Tables, index int, id string) Slot {
	slot := Slot{Index: index, Kind: KindGadget, ID: id}
	gadget, ok := tables.Gadgets[id]
	if !ok {
		return Slot{Index: index, Kind: KindGadget}
	}
	slot.Name, slot.AbilityID = gadget.Name, gadget.Ability
	return slot
}

// spellSlot resolves one of a Mage's slots. The staff is the delivery device
// and the spell is what it casts, so an empty slot on a Mage falls back to the
// staff's own declared spell only in slot zero — that keeps a Mage whose
// loadout has been emptied by a content withdrawal still able to fight.
func spellSlot(tables *tuning.Tables, staff tuning.Weapon, item model.CraftedItem, index int, id string) Slot {
	if id == "" && index == 0 {
		id = staff.Spell
	}
	slot := Slot{Index: index, Kind: KindSpell, ID: id, Item: item}
	spell, ok := tables.Spells[id]
	if !ok {
		return Slot{Index: index, Kind: KindSpell}
	}
	slot.Name, slot.AbilityID, slot.Element = spell.Name, spell.Ability, spell.Element
	return slot
}

// Default is the set a character fights with before it has chosen one: what it
// owns, packed from slot zero. A character's ledger always contains a weapon of
// its class — the starter kit draws one — so the result is a fightable set
// rather than an empty bar, and dropIllegal guarantees it is a legal one.
func Default(tables *tuning.Tables, class model.Class, inventory crafting.Inventory) model.Loadout {
	table := tables.Loadout
	set := model.Loadout{
		Gadgets:   make([]string, table.GadgetSlots),
		Spells:    make([]string, table.SpellSlots),
		Keystones: make([]string, table.KeystoneSlots),
		Version:   tables.Manifest.Version,
	}
	set.Weapon = defaultWeapon(tables, class, inventory)
	if class == model.Gunslinger {
		fill(set.Gadgets, Equippable(tables, class, inventory, KindGadget))
	} else {
		fill(set.Spells, Equippable(tables, class, inventory, KindSpell))
	}
	fill(set.Keystones, Equippable(tables, class, inventory, KindKeystone))
	dropIllegal(tables, class, &set)
	return set
}

// defaultWeapon is the weapon a character falls back to: the first stock row of
// its class it owns, or the class starter when a content withdrawal has left it
// owning none. Being unarmed is not a state any rule allows, so this never
// answers empty while the tables have a starter. It never falls back to a
// crafted instance: the default is the plain configuration, and which crafted
// weapon to carry is a choice the player makes.
func defaultWeapon(tables *tuning.Tables, class model.Class, inventory crafting.Inventory) string {
	if owned := stockWeapons(tables, class, inventory); len(owned) > 0 {
		return owned[0]
	}
	if starter, ok := tables.StarterWeapon(string(class)); ok {
		return starter.ID
	}
	return ""
}

// Resolve carries a saved set onto today's content and guarantees the result is
// legal. Retired IDs follow the retirement chain; anything that no longer
// resolves, that a content change has made illegal, or that the character does
// not own is unequipped rather than dropping the whole set. It reports whether
// the character is owed a respec: a balance patch bumped the manifest, or the
// set itself had to change.
func Resolve(tables *tuning.Tables, class model.Class, inventory crafting.Inventory, saved model.Loadout) (model.Loadout, bool) {
	if saved.Empty() {
		return Default(tables, class, inventory), false
	}
	table := tables.Loadout
	set := model.Loadout{
		Weapon:    resolveWeapon(tables, class, inventory, saved.Weapon),
		Gadgets:   resize(saved.Gadgets, table.GadgetSlots),
		Spells:    resize(saved.Spells, table.SpellSlots),
		Keystones: resize(saved.Keystones, table.KeystoneSlots),
		Version:   tables.Manifest.Version,
	}
	for index, id := range set.Gadgets {
		set.Gadgets[index] = owned(tables, inventory, KindGadget, id)
	}
	for index, id := range set.Spells {
		set.Spells[index] = owned(tables, inventory, KindSpell, id)
	}
	for index, id := range set.Keystones {
		set.Keystones[index] = owned(tables, inventory, KindKeystone, id)
	}
	dropIllegal(tables, class, &set)
	changed := saved.Version != tables.Manifest.Version || !equal(saved, set)
	return set, changed
}

// resolveWeapon carries a saved weapon slot onto today's content. A crafted
// instance the character still owns is kept as it is; a stock row follows its
// retirement chain like any other reference. Anything that no longer resolves,
// belongs to the other class, or is not owned falls back to the default, because
// a withdrawn weapon leaving the character unarmed is a state no rule allows.
func resolveWeapon(tables *tuning.Tables, class model.Class, inventory crafting.Inventory, saved string) string {
	if weapon, _, ok := inventory.Equipped(tables, saved); ok && weapon.Class == string(class) {
		return saved
	}
	if _, isItem := inventory.Item(saved); !isItem {
		if id := owned(tables, inventory, KindWeapon, saved); id != "" && tables.Weapons[id].Class == string(class) {
			return id
		}
	}
	return defaultWeapon(tables, class, inventory)
}

// stockWeapons lists the plain weapon rows of a class the ledger owns, in stable
// order. A row that may only be carried as a crafted instance has no stock
// configuration and never appears here.
func stockWeapons(tables *tuning.Tables, class model.Class, inventory crafting.Inventory) []string {
	ids := make([]string, 0, len(tables.Weapons))
	for id, weapon := range tables.Weapons {
		if weapon.Class == string(class) && !weapon.RequiresCraft && inventory.Ledger.Has(id) {
			ids = append(ids, id)
		}
	}
	sort.Strings(ids)
	return ids
}

// dropIllegal unequips whatever keeps a set from validating, highest slot
// first, so the deterministic casualty is the last thing equipped rather than
// the character's signature. It terminates because each pass empties one slot.
func dropIllegal(tables *tuning.Tables, class model.Class, set *model.Loadout) {
	if class == model.Gunslinger {
		set.Spells = make([]string, tables.Loadout.SpellSlots)
	} else {
		set.Gadgets = make([]string, tables.Loadout.GadgetSlots)
	}
	deduplicate(set.Gadgets)
	deduplicate(set.Spells)
	deduplicate(set.Keystones)
	for index, id := range set.Keystones {
		if id != "" && tables.Keystones[id].Class != string(class) {
			set.Keystones[index] = ""
		}
	}
	for pass := 0; pass < len(set.Spells); pass++ {
		index := worstAffinity(tables, set.Spells)
		if index < 0 {
			return
		}
		set.Spells[index] = ""
	}
}

// worstAffinity finds the highest-indexed slot whose spell does not have the
// same-element company its tier requires, and -1 when every slot is satisfied.
func worstAffinity(tables *tuning.Tables, spells []string) int {
	for index := len(spells) - 1; index >= 0; index-- {
		if spells[index] == "" {
			continue
		}
		if affinityShortfall(tables, spells, index) > 0 {
			return index
		}
	}
	return -1
}

// affinityShortfall is how many more same-element spells the slot's tier needs
// beside it. Zero means the slot is satisfied.
func affinityShortfall(tables *tuning.Tables, spells []string, index int) int {
	spell, ok := tables.Spells[spells[index]]
	if !ok {
		return 0
	}
	required := tables.Loadout.RequiredSameElement(spell.Tier)
	company := 0
	for other, id := range spells {
		if other == index || id == "" {
			continue
		}
		if tables.Spells[id].Element == spell.Element {
			company++
		}
	}
	if company >= required {
		return 0
	}
	return required - company
}

// Validate reports why a requested set may not be equipped, in language a
// player can act on. It is the authority the mutation path runs before it
// commits, and it never consults the world: legality is a property of the set.
func Validate(tables *tuning.Tables, class model.Class, inventory crafting.Inventory, set model.Loadout) error {
	table := tables.Loadout
	ledger := inventory.Ledger
	weapon, _, ok := inventory.Equipped(tables, set.Weapon)
	if !ok {
		// The slot names either a stock row or a crafted instance, so a failure
		// here is one of three things: nothing of that name, an instance another
		// character owns, or a category this one has not unlocked. Naming the row
		// when there is one keeps the message actionable.
		if item, isItem := inventory.Item(set.Weapon); isItem {
			if row, live := tables.Weapons[item.Weapon]; live {
				return fmt.Errorf("you have not unlocked %s", row.Name)
			}
		}
		if row, live := tables.Weapons[set.Weapon]; live {
			if row.RequiresCraft {
				return fmt.Errorf("%s has to be built before it can be carried", row.Name)
			}
			return fmt.Errorf("you have not unlocked %s", row.Name)
		}
		return fmt.Errorf("%q is not a weapon you can equip", set.Weapon)
	}
	if weapon.Class != string(class) {
		return fmt.Errorf("%s is a %s weapon", weapon.Name, weapon.Class)
	}
	if len(set.Gadgets) > table.GadgetSlots {
		return fmt.Errorf("a loadout holds %d gadget slots, not %d", table.GadgetSlots, len(set.Gadgets))
	}
	if len(set.Spells) > table.SpellSlots {
		return fmt.Errorf("a loadout holds %d spell slots, not %d", table.SpellSlots, len(set.Spells))
	}
	if len(set.Keystones) > table.KeystoneSlots {
		return fmt.Errorf("a loadout holds %d keystone slots, not %d", table.KeystoneSlots, len(set.Keystones))
	}
	if class == model.Mage && filled(set.Gadgets) > 0 {
		return fmt.Errorf("a Mage equips spells, not gadgets")
	}
	if class == model.Gunslinger && filled(set.Spells) > 0 {
		return fmt.Errorf("a Gunslinger equips gadgets, not spells")
	}
	if err := checkSlots(ledger, set.Gadgets, "gadget", func(id string) (string, bool) {
		gadget, ok := tables.Gadgets[id]
		return gadget.Name, ok && gadget.Class == string(class)
	}); err != nil {
		return err
	}
	if err := checkSlots(ledger, set.Spells, "spell", func(id string) (string, bool) {
		spell, ok := tables.Spells[id]
		return spell.Name, ok
	}); err != nil {
		return err
	}
	if err := checkSlots(ledger, set.Keystones, "keystone", func(id string) (string, bool) {
		keystone, ok := tables.Keystones[id]
		return keystone.Name, ok && keystone.Class == string(class)
	}); err != nil {
		return err
	}
	for index := range set.Spells {
		if set.Spells[index] == "" {
			continue
		}
		if shortfall := affinityShortfall(tables, set.Spells, index); shortfall > 0 {
			spell := tables.Spells[set.Spells[index]]
			element := tables.Elements[spell.Element].Name
			return fmt.Errorf("%s is tier %d, so it needs %d more %s spell%s beside it",
				spell.Name, spell.Tier, shortfall, element, plural(shortfall))
		}
	}
	return nil
}

// checkSlots rejects unknown, unowned, or duplicated content in one kind of
// slot.
func checkSlots(ledger progression.Ledger, ids []string, kind string, lookup func(string) (string, bool)) error {
	seen := map[string]bool{}
	for _, id := range ids {
		if id == "" {
			continue
		}
		name, ok := lookup(id)
		if !ok {
			return fmt.Errorf("%q is not a %s you can equip", id, kind)
		}
		if !ledger.Has(id) {
			return fmt.Errorf("you have not unlocked %s", name)
		}
		if seen[id] {
			return fmt.Errorf("%s is already equipped in another slot", name)
		}
		seen[id] = true
	}
	return nil
}

// Equippable lists the content of a slot kind a character may choose from, in
// stable order: the live rows of the right class that its ledger owns. Owning
// more options improves preparation, never the power carried into one fight, so
// this is the only place the ledger narrows.
func Equippable(tables *tuning.Tables, class model.Class, inventory crafting.Inventory, kind string) []string {
	ledger := inventory.Ledger
	var ids []string
	switch kind {
	case KindWeapon:
		// Stock rows first, then the crafted instances of them, so the plain
		// configuration is the deterministic default and a crafted weapon is
		// something the player picks on purpose.
		stock := stockWeapons(tables, class, inventory)
		crafted := make([]string, 0, len(inventory.Items))
		for _, item := range inventory.Items {
			if weapon, ok := tables.Weapons[item.Weapon]; ok && weapon.Class == string(class) && ledger.Has(item.Weapon) {
				crafted = append(crafted, item.ID)
			}
		}
		sort.Strings(crafted)
		return append(stock, crafted...)
	case KindGadget:
		for id, gadget := range tables.Gadgets {
			if gadget.Class == string(class) && ledger.Has(id) {
				ids = append(ids, id)
			}
		}
	case KindSpell:
		if class != model.Mage {
			return nil
		}
		for id := range tables.Spells {
			if ledger.Has(id) {
				ids = append(ids, id)
			}
		}
	case KindKeystone:
		for id, keystone := range tables.Keystones {
			if keystone.Class == string(class) && ledger.Has(id) {
				ids = append(ids, id)
			}
		}
	}
	sort.Strings(ids)
	return ids
}

// owned carries one saved slot onto today's content: the retirement chain is
// followed, and the result is kept only if the character actually owns it. A
// slot holding something the ledger does not have empties rather than granting
// it — the ledger is the authority on ownership, not the saved set.
func owned(tables *tuning.Tables, inventory crafting.Inventory, kind, id string) string {
	if id == "" {
		return ""
	}
	resolved, ok := tables.Resolve(kind, id)
	// A refunded retirement has no live replacement, so the slot empties. The
	// refund itself is owed against the material ledger, not the slot.
	if !ok || !inventory.Ledger.Has(resolved.ID) {
		return ""
	}
	return resolved.ID
}

// deduplicate empties the later of any two slots holding the same content, so
// one spell cannot pad its own affinity requirement.
func deduplicate(ids []string) {
	seen := map[string]bool{}
	for index, id := range ids {
		if id == "" {
			continue
		}
		if seen[id] {
			ids[index] = ""
			continue
		}
		seen[id] = true
	}
}

func resize(ids []string, size int) []string {
	sized := make([]string, size)
	copy(sized, ids)
	return sized
}

func fill(slots []string, ids []string) {
	for index := range slots {
		if index >= len(ids) {
			return
		}
		slots[index] = ids[index]
	}
}

func at(ids []string, index int) string {
	if index >= len(ids) {
		return ""
	}
	return ids[index]
}

func filled(ids []string) int {
	count := 0
	for _, id := range ids {
		if id != "" {
			count++
		}
	}
	return count
}

// equal compares two sets by what they equip, ignoring how many trailing empty
// slots each one happens to carry — a record saved before the slot count grew
// equips exactly what a padded one does.
func equal(a, b model.Loadout) bool {
	return a.Weapon == b.Weapon &&
		key(a.Gadgets) == key(b.Gadgets) &&
		key(a.Spells) == key(b.Spells) &&
		key(a.Keystones) == key(b.Keystones)
}

func key(ids []string) string {
	end := len(ids)
	for end > 0 && ids[end-1] == "" {
		end--
	}
	return strings.Join(ids[:end], "\x00")
}

func plural(count int) string {
	if count == 1 {
		return ""
	}
	return "s"
}

File: web/src/game/loadout.ts
// The client's view of the equipped set. It mirrors server/internal/loadout so
// the menu can lay out slots, offer only equippable content, and explain a
// rejection before the round trip — but the server remains the authority: a
// commit is only real once its Loadout reply confirms it.
import { elements, gadgets, loadoutTable, spells, weapons } from "../tuning";
import type { CharacterClass, CraftedItem, LoadoutSet } from "../types";
import { equippableItems, itemLabel } from "./crafting";

export type SlotKind = "weapon" | "gadget" | "spell";

/** One selectable position, mirroring server/internal/loadout.Slot. `abilityId`
 * is what the use button performs, so nothing downstream has to know which
 * table the slot's content came from. */
export interface Slot { index: number; kind: SlotKind; id: string; name: string; element: string; abilityId: string }

/**
 * Resolves an equipped weapon reference to the row it fights with. The slot
 * holds either a stock weapon row or a crafted instance of one; the server
 * resolves the same two cases, so the menu never offers something it would
 * refuse.
 */
export function equippedWeapon(id: string, items: readonly CraftedItem[]): { row: string; item?: CraftedItem } | undefined {
  const item = items.find((owned) => owned.id === id);
  if (item) return weapons[item.weapon] ? { row: item.weapon, item } : undefined;
  // A row the economy withholds has no stock configuration: it can only be
  // carried as an instance that was actually built.
  return weapons[id] && !weapons[id]!.requires_craft ? { row: id } : undefined;
}

/** Selectable action-bar slots, bound to 1–6. One width for both classes. */
export const barSlots = loadoutTable.spell_slots;

/** How many other spells of the same element a tier-N spell needs beside it. */
export function requiredSameElement(tier: number): number {
  return tier <= 1 ? 0 : (tier - 1) * loadoutTable.affinity.same_element_per_tier;
}

/** The equipped set laid out in binding order. */
export function bar(characterClass: CharacterClass, set: LoadoutSet, items: readonly CraftedItem[] = []): Slot[] {
  const equipped = equippedWeapon(set.weapon, items);
  const weapon = equipped ? weapons[equipped.row] : undefined;
  if (characterClass === "gunslinger") {
    const slots: Slot[] = [{ index: 0, kind: "weapon", id: weapon ? set.weapon : "", name: weapon?.name ?? "", element: "", abilityId: weapon?.ability ?? "" }];
    for (let index = 0; index < loadoutTable.gadget_slots; index++) {
      const gadget = gadgets[set.gadgets[index] ?? ""];
      slots.push({ index: slots.length, kind: "gadget", id: gadget ? set.gadgets[index]! : "", name: gadget?.name ?? "", element: "", abilityId: gadget?.ability ?? "" });
    }
    return slots;
  }
  const staff = weapon;
  const slots: Slot[] = [];
  for (let index = 0; index < loadoutTable.spell_slots; index++) {
    // A Mage's empty first slot falls back to the staff's own spell, so a set
    // emptied by a content withdrawal can still fight.
    const id = set.spells[index] || (index === 0 ? staff?.spell ?? "" : "");
    const spell = spells[id];
    slots.push({ index, kind: "spell", id: spell ? id : "", name: spell?.name ?? "", element: spell?.element ?? "", abilityId: spell?.ability ?? "" });
  }
  return slots;
}

/**
 * The set a character fights with before the server's authoritative one
 * arrives: what its ledger owns, packed from slot zero. The server resolves the
 * same shape, so the menu is not showing content the world would refuse.
 */
export function defaultLoadout(characterClass: CharacterClass, ledger: Ledger): LoadoutSet {
  const set: LoadoutSet = {
    // Stock rows only: the default is the plain configuration, and which crafted
    // weapon to carry is a choice the player makes.
    weapon: equippable(characterClass, ledger, "weapon")[0] ?? "",
    gadgets: new Array<string>(loadoutTable.gadget_slots).fill(""),
    spells: new Array<string>(loadoutTable.spell_slots).fill(""),
  };
  const kind: SlotKind = characterClass === "gunslinger" ? "gadget" : "spell";
  const target = characterClass === "gunslinger" ? set.gadgets : set.spells;
  equippable(characterClass, ledger, kind).slice(0, target.length).forEach((id, index) => { target[index] = id; });
  dropIllegal(set.spells);
  return set;
}

/**
 * Unequips whatever keeps a set from validating, highest slot first, exactly as
 * the server's resolve does: packing every owned spell into the bar will often
 * break affinity, and the deterministic casualty is the last thing equipped
 * rather than the character's signature.
 */
function dropIllegal(spells: string[]): void {
  for (let pass = 0; pass < spells.length; pass++) {
    let worst = -1;
    for (let index = spells.length - 1; index >= 0 && worst < 0; index--) {
      if (spells[index] && affinityShortfall(spells, index) > 0) worst = index;
    }
    if (worst < 0) return;
    spells[worst] = "";
  }
}

/** The flat permanent unlock ledger, as the client holds it. */
export type Ledger = ReadonlySet<string>;

export function ledgerOf(unlocks: readonly string[]): Ledger { return new Set(unlocks); }

/**
 * Content of a slot kind the character may choose from, in stable order: the
 * live rows of its class that its ledger owns. The server enforces the same
 * intersection — hiding an option is never what stops it being equipped.
 */
export function equippable(characterClass: CharacterClass, ledger: Ledger, kind: SlotKind, items: readonly CraftedItem[] = []): string[] {
  const owns = (id: string) => ledger.has(id);
  if (kind === "weapon") {
    // Stock rows first, then the crafted instances of them, so the plain
    // configuration stays the deterministic default.
    const stock = Object.keys(weapons).filter((id) => weapons[id]!.class === characterClass && !weapons[id]!.requires_craft && owns(id)).sort();
    return [...stock, ...equippableItems(characterClass, ledger, items).map((item) => item.id)];
  }
  if (kind === "gadget") return Object.keys(gadgets).filter((id) => gadgets[id]!.class === characterClass && owns(id)).sort();
  return characterClass === "mage" ? Object.keys(spells).filter(owns).sort() : [];
}

/**
 * Content of a slot kind the character has not unlocked yet, with the level
 * that grants it. The menu shows these beside what is equippable, disabled: a
 * player who cannot see that a gadget exists has no reason to believe the slot
 * will ever fill, and progression is only motivating if it is legible.
 */
export function locked(characterClass: CharacterClass, ledger: Ledger, kind: SlotKind): LockedContent[] {
  const rows: Record<string, { name: string; unlock_level: number; class?: CharacterClass }> =
    kind === "weapon" ? weapons : kind === "gadget" ? gadgets : (characterClass === "mage" ? spells : {});
  return Object.keys(rows)
    .filter((id) => !ledger.has(id) && (rows[id]!.class ?? characterClass) === characterClass)
    .map((id) => ({ id, name: rows[id]!.name, level: rows[id]!.unlock_level }))
    .sort((left, right) => left.level - right.level || left.id.localeCompare(right.id));
}

export interface LockedContent { id: string; name: string; level: number }

/** Display name of equippable content, whatever kind it is. */
export function contentName(kind: SlotKind, id: string, items: readonly CraftedItem[] = []): string {
  if (kind === "weapon") {
    const item = items.find((owned) => owned.id === id);
    if (item) return itemLabel(item);
    return weapons[id]?.name ?? id;
  }
  if (kind === "gadget") return gadgets[id]?.name ?? id;
  return spells[id]?.name ?? id;
}

/** How many more same-element spells a slot's tier needs. Zero when satisfied. */
export function affinityShortfall(equipped: string[], index: number): number {
  const spell = spells[equipped[index] ?? ""];
  if (!spell) return 0;
  const company = equipped.filter((id, other) => other !== index && id && spells[id]?.element === spell.element).length;
  return Math.max(0, requiredSameElement(spell.tier) - company);
}

/**
 * Why a set may not be committed, or undefined when it is legal. This mirrors
 * the server's rules so the menu can refuse before spending a round trip; the
 * server still validates and its answer wins.
 */
export function loadoutProblem(characterClass: CharacterClass, ledger: Ledger, set: LoadoutSet, items: readonly CraftedItem[] = []): string | undefined {
  const held = equippedWeapon(set.weapon, items);
  const weapon = held ? weapons[held.row] : undefined;
  if (!weapon || !held) {
    const withheld = weapons[set.weapon];
    if (withheld?.requires_craft) return `${withheld.name} has to be built before it can be carried.`;
    return "Choose a weapon.";
  }
  if (weapon.class !== characterClass) return `${weapon.name} is a ${weapon.class} weapon.`;
  if (!ledger.has(held.row)) return `You have not unlocked ${weapon.name}.`;
  const equipped = characterClass === "gunslinger" ? set.gadgets : set.spells;
  const kind: SlotKind = characterClass === "gunslinger" ? "gadget" : "spell";
  const seen = new Set<string>();
  for (const id of equipped) {
    if (!id) continue;
    if (!ledger.has(id)) return `You have not unlocked ${contentName(kind, id)}.`;
    if (seen.has(id)) return `${contentName(kind, id)} is already equipped in another slot.`;
    seen.add(id);
  }
  for (let index = 0; index < set.spells.length; index++) {
    const shortfall = affinityShortfall(set.spells, index);
    if (!shortfall) continue;
    const spell = spells[set.spells[index]!]!;
    const element = elements[spell.element]?.name ?? spell.element;
    return `${spell.name} is tier ${spell.tier}, so it needs ${shortfall} more ${element} spell${shortfall === 1 ? "" : "s"} beside it.`;
  }
  return undefined;
}
